### ISSUE TITLE
Only load a image to view

### ISSUE THREAD

**Issue by mdtuyen** [bumptech/glide#584]
I use Glide and Picasso to load image to a View (have imageview), this view will be convert to bittmap to make icon for a marker in googlemap:

With picasso:
```
        Picasso.with(mActivity).load(Uri.parse(feedsModel.getImages().getThumbnail().getUrl())).resize(50, 50).into(image, new Callback() {
            @Override
            public void onSuccess() {
                Marker marker = mMap.addMarker(new MarkerOptions()
                        .position(mLatLng)
                        .icon(BitmapDescriptorFactory.fromBitmap(createDrawableFromView(mActivity, markerView))));
                markers.put(marker, feedsModel);

            }
}
```

It work fine when i call it many time for a list location

However when i use Glide:
```
        Glide.with(mActivity)
                .load(feedsModel.getImages().getThumbnail().getUrl())
                .asBitmap()
                .into(new BitmapImageViewTarget(image){
                    @Override
                    public void onResourceReady(Bitmap resource, GlideAnimation<? super Bitmap> glideAnimation) {
                        super.onResourceReady(resource, glideAnimation);
                        image.setImageBitmap(resource);
                        Marker marker = mMap.addMarker(new MarkerOptions()
                                .position(mLatLng)
                                .icon(BitmapDescriptorFactory.fromBitmap(createDrawableFromView(mActivity, markerView))));
                        markers.put(marker, feedsModel);
                    }
                });
```

It only work with one point, with a list point in a loop, only one maker is created, it is last marker in list. (Result the same with SimpleTarget, RequestListener)

Please explain for me about this issue. What do i have to make it work the same picasso.

**Comment by TWiStErRob** [comment]
Hi, welcome to Glide/GitHub. Please look into how to [format your comments](https://guides.github.com/features/mastering-markdown/) so they're readable, most notably wrap your code blocks with ``` before and after, see the linked Syntax Highlight section.

In general if you have one `ImageView` and you load an image into it, it will be displayed, if you load another image, then the new image will be displayed; that's just how you would expect it to work by common sense. Now, you're abusing this by loading multiple images into the same view and expecting them all to be loaded one after the other. It looks like while Picasso will go through with the full load, that's why it works for multiple markers, Glide cancels any previous request before starting the new load. Think of Glide's viewpoint like this: "Why download a 5MB image, if another 2kB image will be displayed instead?".

You can make this work without the `createDrawableFromView` hack, which is causing the problem:
```java
Glide
	.with(mActivity)
	.load(feedsModel.getImages().getThumbnail().getUrl())
	.asBitmap()
	//.fitCenter() or .centerCrop() depending on what was the android:scaleType on the ImageView
	.into(new SimpleTarget<Bitmap>(50, 50) {
		@Override public void onResourceReady(Bitmap resource, GlideAnimation<? super Bitmap> glideAnimation) {
			Marker marker = mMap.addMarker(new MarkerOptions()
					.position(mLatLng)
					.icon(BitmapDescriptorFactory.fromBitmap(resource))
			);
			markers.put(marker, feedsModel);
		}
	})
;
```

**Comment by mdtuyen** [comment]
Thanks for your reply, I tried your solution and result no change. createDrawableFromView  function help me custom maker because beside image, maker have other info

**Comment by TWiStErRob** [comment]
You should use a [custom transformation](https://github.com/bumptech/glide/wiki/Transformations#custom-transformations) for adding extra information to a Bitmap.

Would it be possible to create a mini-project demonstrating your issue or do you have an open-source repo? There may be more to your issue than seen here. For example the `markers` map may be used too early, not sure how you synchronize on that; all the `onResourceReady` methods are called on a background thread, so even though all the `into()` calls are finished, the resources are not ready at the end of that loop.

**Comment by mdtuyen** [comment]
This is my code:
```for (FeedsModel feedsModel : ((FeedData)serializable).getFeedsModels()) {
	if (feedsModel.getLocation() != null) {
		mLatLng = new LatLng(
			Double.parseDouble(feedsModel.getLocation().getLatitude()), 
			Double.parseDouble(feedsModel.getLocation().getLongitude()));
		setUpMap(feedsModel);
	}
};```

and setUpMap:
```Glide.with(mActivity)
	.load(feedsModel.getImages().getThumbnail().getUrl())
	.asBitmap()
	.into(new SimpleTarget<Bitmap>(50, 50) {
		@Override
		public void onResourceReady(Bitmap resource, GlideAnimation<? super Bitmap> glideAnimation) {
			Marker marker = mMap.addMarker(new MarkerOptions()
							.position(mLatLng)
							.icon(BitmapDescriptorFactory.fromBitmap(resource))
			);
			markers.put(marker, feedsModel);
		}
	});```

**Comment by TWiStErRob** [comment]
And where do you use `markers`?

You should also try adding a [`.listener()`](https://github.com/bumptech/glide/wiki/Debugging-and-Error-Handling#requestlistener) for these loads and logging all method calls (`listener.onException`, `listener.onResourceReady`, `target.onResourceReady`) with arguments to see which is called and when.

**Comment by TWiStErRob** [comment]
You need new lines after and before ``` (e.g. `Glide` is missing from your last comment), you can edit your previous comments in the upper right corner of them via the pencil icon.

**Comment by mdtuyen** [comment]
Thanks for your rely, i use "markers" when i click to marker on google map. I tried many time and only one marker appear so that **(listener.onException, listener.onResourceReady, target.onResourceReady) ** is not proplem, I think but i will test it.

**Comment by TWiStErRob** [comment]
Does `Marker` have a proper `hashCode` implementation? Maybe their hashes are the same or `Comparator/Comparable` returns "equal" and the markers `Map` contains only 1 item... though then you'd have the same problem with Picasso too.

I'm off for now; I'll come back in 9 hours, let me know what you find out.

**Comment by mdtuyen** [comment]
This problem no occur with picasso, i do not modify Marker of Google.
Only The line load image to marker is difference with picasso solution (however picasso load image is low, so that i want have a solution with glide)

**Comment by mdtuyen** [comment]
Thanks for your help,
Good night

See you tomorow ! :)

**Comment by TWiStErRob** [comment]
I had an insight: Convert mLatLng to a local inside setUpMap, otherwise the for loop will quickly go through leaving the last entry there. And then the marker's position is probably the hashCode of it, which means that when Glide finishes async it'll create the same marker many times with different images at the same location.

**Comment by mdtuyen** [comment]
I thank it too, so that i have changed as:
```
                        for (FeedsModel feedsModel : ((FeedData)serializable).getFeedsModels()) {
                            if (feedsModel.getLocation() != null) {
                                new AsyncTask<FeedsModel, Void, Void>(){
                                    @Override
                                    protected Void doInBackground(FeedsModel... params) {
                                        mLatLng = new LatLng(Double.parseDouble(params[0].getLocation().getLatitude()), Double.parseDouble(params[0].getLocation().getLongitude()));
                                        setUpMap(params[0]);
                                        return null;
                                    }
                                }.execute(feedsModel);
                            }
                        }
```
:(   no fine. error

**Comment by MIkeeJY** [comment]
Why not try  imageloader by nostra13？

**Comment by TWiStErRob** [comment]
@MIkeeJY that won't fix his problem, because it's just a Java programming issue, from what I see now.

@mdtuyen The problem with both the original and your last try is that you're using a field to store information temporarily, but once you have an anonymous inner class which executes asynchrornously that just cleanly leads to bugs. Imagine the execution order of statements with list of 3 items:
```java
for (models[i] : models) {
    mLatLng = models[0].location
    setupMap(models[0])
    mLatLng = models[1].location // overwriting models[0].location
    setupMap(models[1])
    mLatLng = models[2].location // overwriting models[1].location
    setupMap(models[2])
}
// from here on mLatLng has the value of models[2].location
// Glide schedules and processes your load()...into() requests asynchronously
// notice that execution is not even linear, you don't know the order in which they'll execute
SimpleTarget.onResourceReady(bitmapForModels[2])
SimpleTarget.onResourceReady(bitmapForModels[0])
SimpleTarget.onResourceReady(bitmapForModels[1])
```

In your last comment you tried to fix this with `AsyncTask`, but with that all you achieve is that the inside of the for loop is executed on a background thread instead of the main thread. Also the order of execution may become random even for `mLatLng = models[i].location` if AsyncTasks are executed in parallel. This leads to a similar execution order of statements as above with the exception that after the for loop `mLatLng` has 4 possible values (unassigned, `models[0..2].location`), and it's impossible to predict which one.

Let's go through this one (which hopefully solves your issue):
```java
for (FeedsModel feedsModel : ((FeedData)serializable).getFeedsModels()) {
    if (feedsModel.getLocation() != null) {
        setupMap(feedsModel); // method is called syncronously
    }
}

void setupMap(final FeedsModel feedsModel) {
    Glide
            .with(mActivity)
            .load(feedsModel.getImages().getThumbnail().getUrl())
            .asBitmap()
            .into(new SimpleTarget<Bitmap>(50, 50) {
                // from here on it'll be executed later, asynchronously
                // feedsModel is saved in "closure"
                // and will be accessible inside this anonymous inner class
                @Override
                public void onResourceReady(Bitmap resource, GlideAnimation<? super Bitmap> glideAnimation) {
                    // here feedsModel is the same as the one used for loading the image just a few lines above
                    // but this code is executed much later and will be executed 3 times (like the above ones) in unknown order
                    // but order doesn't matter because the location is converted here from the right feedsModel
                    LatLng location = new LatLng( // notice the local variable, not a field
                            Double.parseDouble(feedsModel.getLocation().getLatitude()),
                            Double.parseDouble(feedsModel.getLocation().getLongitude())
                    );
                    Marker marker = mMap.addMarker(new MarkerOptions()
                            .position(location)
                            .icon(BitmapDescriptorFactory.fromBitmap(resource))
                    );
                    markers.put(marker, feedsModel);
                }
            })
    ;
    // anything here will still execute synchronously inside the for loop of the calling method
}
```

Sidenote: if you need the location inside `setupMap`, you can move it outside the inner class, the closure will take care of it:
```java
void setupMap(final FeedsModel feedsModel) {
    final LatLng location = new LatLng( // notice the local variable, not a field, passed to SimpleTarget with a closure (final)
            Double.parseDouble(feedsModel.getLocation().getLatitude()),
            Double.parseDouble(feedsModel.getLocation().getLongitude())
    );
    Glide.....into(new SimpleTarget<Bitmap>(50, 50) {
        @Override public void onResourceReady(Bitmap resource, GlideAnimation<? super Bitmap> glideAnimation) {
            Marker marker = mMap.addMarker(new MarkerOptions()
                    .position(location) // referenced from setupMap method
                    .icon(BitmapDescriptorFactory.fromBitmap(resource))
            );
            // ...
        }
    });
    // use variable location here as well for anything else
}
```

**Comment by mdtuyen** [comment]
@TWiStErRob Thanks you so much. With your help i have my solution, it load image very fast (although slightly longer than picasso)
```
                        markerView = LayoutInflater.from(mActivity).inflate(R.layout.custom_marker_layout, null);
                        ImageView image = (ImageView) markerView.findViewById(R.id.image);
                        image.setImageBitmap(resource);
                        Marker marker = mMap.addMarker(new MarkerOptions()
                                        .position(location)
                                        .icon(BitmapDescriptorFactory.fromBitmap(createDrawableFromView(mActivity, markerView)))
                        );
                        markers.put(marker, feedsModel);
```

**Comment by TWiStErRob** [comment]
If you rewrite `createDrawableFromView` to be a `BitmapTransformation` it should be much faster; feel free to drop me an email if you need help with it.

**Comment by sjudd** [comment]
Glad you guys got this figured out. Thanks @TWiStErRob!

**Comment by ArisChoice** [comment]
how to use this with circular image view.

**Comment by TWiStErRob** [comment]
@ArisChoice where is the circular view? The point of this that there's no ImageView, but a Maps marker to load into. See [Compatibility](https://github.com/bumptech/glide#compatibility) for more info on circle-views. The general advice is to use a [transformation](https://github.com/wasabeef/glide-transformations/blob/master/transformations/src/main/java/jp/wasabeef/glide/transformations/CropCircleTransformation.java) that removes pixels from your `Bitmap`.

If the above doesn't provide sufficient information, please open a new issue with all the details filled in, so we can give you a solution.
